File: Python3/85. Maximal Rectangle.py
# 216ms, 82.08%; 15MB, 6.25%
class Solution:
    def maximalRectangle(self, matrix: list) -> int:
        if not matrix: return 0
        row, col = len(matrix), len(matrix[0])
        // the consecutive '1's begin at index left[j] and end at index right[j](not included)
        left, right, height = [0] * col, [col] * col, [0] * col
        res = 0
        for i in range(row):
            cur_left, cur_right = 0, col
            for j in range(col):
                if matrix[i][j] == '1':
                    left[j] = max(left[j], cur_left)
                    height[j] += 1
                else:
                    left[j] = height[j] = 0
                    cur_left = j + 1
            for j in range(col - 1, -1, -1):
                if matrix[i][j] == '1':
                    right[j] = min(right[j], cur_right)
                    res = max(res, height[j] * (right[j] - left[j]))
                else:
                    right[j] = col
                    cur_right = j
        return res

File: 1 - 200/85 Maximal Rectangle.cc
// 20ms, 94.19%; 10.4MB, 100.00%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maximalRectangle(const vector<vector<char>> &matrix) // dp
    {
        if (matrix.empty()) return 0;
        int row = matrix.size(), col = matrix[0].size();
        vector<int> left(col), right(col, col), height(col);
        int res = 0;
        for (int i = 0; i < row; ++i) {
            int cur_left = 0, cur_right = col;
            for (int j = 0; j < col; ++j) {
                if (matrix[i][j] == '1') {
                    left[j] = max(left[j], cur_left);
                    height[j] += 1;
                } else {
                    left[j] = 0, cur_left = j + 1;
                    height[j] = 0;
                }
            }
            for (int j = col - 1; j > -1; --j) {
                if (matrix[i][j] == '1') {
                    right[j] = min(right[j], cur_right);
                    res = max(res, height[j] * (right[j] - left[j]));
                } else {
                    right[j] = col, cur_right = j;
                }
            }
        }
        return res;
    }
};
